### PR TITLE
Add a golden test for duplicate certificates

### LINKED ISSUE #4566: Fail on duplicate certificates
**Internal/External**
*Internal*

**Area**
*Other*   Any other topic (Delegation, Ranking, ...).

**Summary**
In Babbage transaction, the order of certificates is retained, even when the certificates are repeated. I can eg. build a tx that contain following certs in this order

1. stake address registration cert
2. deregistration cert for the stake address above
3. stake address registration cert from 1.
4. deregistration cert from 2.
5. stake address registration cert from 1.

The transaction can be submitted and the stake address is registered.

Certs in Babbage tx:

```json
    "certificates": [
        {
            "stake address registration": {
                "keyHash": "e24c2f4946e87eae96ace9c362b3c6a0d40a0c3a04592a591824180a"
            }
        },
        {
            "stake address deregistration": {
                "keyHash": "e24c2f4946e87eae96ace9c362b3c6a0d40a0c3a04592a591824180a"
            }
        },
        {
            "stake address registration": {
                "keyHash": "e24c2f4946e87eae96ace9c362b3c6a0d40a0c3a04592a591824180a"
            }
        },
        {
            "stake address deregistration": {
                "keyHash": "e24c2f4946e87eae96ace9c362b3c6a0d40a0c3a04592a591824180a"
            }
        },
        {
            "stake address registration": {
                "keyHash": "e24c2f4946e87eae96ace9c362b3c6a0d40a0c3a04592a591824180a"
            }
        }
    ],
```

However in Conway tx, the repeated certs are stripped. Therefore when submitting the tx with certificates order listed above, the stake address is not registered.

Certs in Conway tx:

```json
    "certificates": [
        {
            "Stake address registration": {
                "stake credential": {
                    "keyHash": "4f2792103fe76cb96a0c5fb89dcd3f36441a2f9c697dc783ed915039"
                }
            }
        },
        {
            "Stake address deregistration": {
                "stake credential": {
                    "keyHash": "4f2792103fe76cb96a0c5fb89dcd3f36441a2f9c697dc783ed915039"
                }
            }
        }
    ],

```

**Steps to reproduce**
1. build a tx

> cardano-cli conway transaction build --tx-in "57e67b2775839788da91936645f98caac34a157a2f01dfc3e5334659160758cd#0" --certificate-file "test_addr_registration_certificate_order[build-submit_cli]_ci0_ykz_addr0_stake_reg.cert" --certificate-file "test_addr_registration_certificate_order[build-submit_cli]_ci0_ykz_addr0_stake_dereg.cert" --certificate-file "test_addr_registration_certificate_order[build-submit_cli]_ci0_ykz_addr0_stake_reg.cert" --certificate-file "test_addr_registration_certificate_order[build-submit_cli]_ci0_ykz_addr0_stake_dereg.cert" --certificate-file "test_addr_registration_certificate_order[build-submit_cli]_ci0_ykz_addr0_stake_reg.cert" --change-address addr_test1qz8el0prsndp9y00tvv4aykmjhrteg450e3f72v6n8rxl6ymn7p2shwt24atpgw7fzshqdjjl7c9cfr3skrkjq5n098sf5dk3h --witness-override 2 --out-file "test_addr_registration_certificate_order[build-submit_cli]_ci0_ykz_tx.body" --testnet-magic 42

2. sign and submit the tx, see that the stake address was not registered
3. View the tx, see that repeated certificates were stripped

**Expected behavior**

**System info (please complete the following information):**
- OS Name: Fedora 37
- Node / CLI version:
> cardano-node 8.9.0 - linux-x86_64 - ghc-8.10
git rev 0d98405a60d57e1c8e13406d51cce0e34356bd64

**Screenshots and attachments**
Conway and Babbage transactions: [tx_certificates_issue.tar.gz](https://github.com/IntersectMBO/cardano-api/files/14599965/tx_certificates_issue.tar.gz)

### PR DESCRIPTION
**Pull request by Soupstraw**
> [!NOTE]
> The base branch here (`leios-prototype`) is the state before the recreation;
> this PR is intended to merge by fast-forwarding `leios-prototype` to the head
> of `leios-prototype-remake`. The diff is huge because it includes the
> snapshot-merge to upstream `main`. For focused review:
> - Recreation against current main: https://github.com/IntersectMBO/ouroboros-network/compare/main...leios-prototype-remake
> - Per-commit first-parent view: `git log --first-parent ebc8eac0a8..leios-prototype-remake`

> [!IMPORTANT]
> Cross-repo: this branch is consumed by `ouroboros-consensus`'s
> `leios-prototype-remake` (PR IntersectMBO/ouroboros-consensus#2041), where
> the Leios threadnet test is currently green (20 QC iters / 78 s) with this
> branch pinned via SRP and the BearerBytes + Reception adapters wired in.

## Done

- [x] **Snapshot-merge onto upstream `main`** (`ebc8eac0a8`) — the
      `leios-prototype-remake` ref is reset to upstream `main`'s tree while
      keeping the prototype's historic commits reachable via the merge
      commit's first parent.
- [x] **Demo-tuning: `BearerBytes` class replaces `dataSize`** (`55c16cf687`)
      Drop the `dataSize :: bytes -> Word` field from `ProtocolSizeLimits`;
      replace with a `class BearerBytes bytes where bearerBytesSize :: bytes
      -> Word` and default instances for `BS.ByteString`, `BL.ByteString`,
      `[Char]`, `AnyMessage msg`. The seven `byteLimits*` codec helpers no
      longer take a `(bytes -> Word)` argument; call sites drop the
      `(fromIntegral . LBS.length)` boilerplate. `Driver.Limits` gains a
      `BearerBytes bytes` constraint on the public entry points
      (`runPeerWithLimits` and friends).
- [x] **Demo-tuning: arrival-time `Reception` on `recv` + `TraceRecvMsg`**
      (`319743abd0`) — bearer-side recv now returns `Maybe (Reception a)`
      where `Reception a = MkReception !(IntMap Time) !a` records per-chunk
      arrival times keyed by starting-byte offset, and that arrival time is
      threaded all the way through the decoder to surface on `TraceRecvMsg ::
      Maybe Time -> AnyMessage ps -> TraceSendRecv ps` (with `TraceSendMsg ::
      Time -> AnyMessage ps -> ...` for symmetry). The mux demuxer pairs each
      SDU's read time with its offset into the ingress queue
      (`IngressQueueVal Int64 (IntMap Time) Builder`). The driver's decoder
      threads `(IntMap Time, Word rsz, Word sz)` so the next-message arrival
      time isn't lost when a message boundary lands inside a chunk
      (`runDecoderWithChannel_DecodeDone` helper, `mbEq <|> mbTm` fallback).
      `Reception` is re-exported from `Network.Mux` and `Ouroboros.Network.Mux`;
      `MonadMonotonicTime m` propagates to a handful of test entry points.
      *Squashes the prototype's original `Reception` drop together with the
      subsequent `TraceRecvMsg` consumer-side wiring + the four `runDecoder*`
      bugfixes that followed.*
- [x] **Demo-tuning: field names on `Reception`** (`3afd70f505`) — name the
      `MkReception` fields `receivalTimes` / `received` for ergonomic access.

## Still TODO

- [ ] `TCP_NOTSENT_LOWAT` + `TCP_NODELAY` on socket bearer set on the
      `mw/leios-prototype` branch — can be rebased.
- [ ] Open question: can these add-ons be PRed straight upstream into
      `ouroboros-network` master, eliminating the need for a fork branch?
      The two demo-tuning patches are self-contained and don't require any
      Leios protocol context, so they could land independently.

## Intentional differences from the original prototype

The Leios mini-protocols (`LeiosNotify`, `LeiosFetch`) live in `ouroboros-consensus`
rather than here — easier iteration in one repo. That's why this branch carries
only the bearer-layer demo-tuning (arrival times, `BearerBytes`) but no Leios
protocol content.

## Merge plan

Fast-forward `leios-prototype` to the head of this branch (after the
TODOs above are resolved).

### CODE REVIEW

File: eras/dijkstra/impl/testlib/Test/Cardano/Ledger/Dijkstra/Binary/Golden.hs [hunk 1]
```diff
@@ -1,27 +1,83 @@
...
-    expectDecoderFailure @(Redeemers era)
+    expectDecoderFailureAnn @(Redeemers era)
       (eraProtVerLow @era)
       listRedeemersEnc
       ( DecoderErrorDeserialiseFailure
           "Annotator (MemoBytes (RedeemersRaw DijkstraEra))"
           (DeserialiseFailure 0 "List encoding of redeemers not supported starting with PV 12")
       )
+
+duplicateCertsTx :: forall era. DijkstraEraTest era => Version -> Enc
```
Comment (lehins) [inline eras/dijkstra/impl/testlib/Test/Cardano/Ledger/Dijkstra/Binary/Golden.hs:49]: I think it would be an awesome addition if we moved this example to shelley era and added it as a positive test, where:
* In Shelley till Babbage we successfully decode such transaction, but only one certificate is preserved
* In Conway only one certificate is preserved

**Review by lehins — APPROVED** [review]
This looks great. Only made a suggestion which would take this test to a whole another level.

File: eras/dijkstra/impl/testlib/Test/Cardano/Ledger/Dijkstra/Binary/Golden.hs (hunk 1 above)
Comment (Soupstraw) [reply eras/dijkstra/impl/testlib/Test/Cardano/Ledger/Dijkstra/Binary/Golden.hs:49]: I made a separate issue for this, I'd rather merge this PR because there's another PR sitting on top of it.
